src/ConstraintSolver.jl

```diff
@@ -83,26 +83,60 @@ struct NotEqualSet{T} <: MOI.AbstractScalarSet
     value :: T
 end
 
-mutable struct LinearVariables
+mutable struct LinearCombination{T <: Real}
     indices             :: Vector{Int}
-    coeffs              :: Vector{Int}
+    coeffs              :: Vector{T}
 end
 
-mutable struct LinearConstraint <: Constraint
+mutable struct LinearConstraint{T <: Real} <: Constraint
     idx                 :: Int
     fct                 :: Function
```

Comment:
This may be for another PR but do you need to store the function here?

Reply:
Currently every constraint has a function attached to it to call the corresponding constraint functions (pruning and feasibility check) later. 
- Is this unusual or not advised?
- What would be other reasonable ways to link the constraint type to the functions then?
Open for suggestions :)

Reply:
Ok I see thanks for clarifying. So it's a bit unusual, but also expensive. The field of the struct is not concretely typed, just like the elements of a `Vector{Real}`. This yields to boxing and dynamic dispatch every time (slower). I think I will think of a proposal in a separate PR / issue

Reply:
Hmm okay makes sense. Maybe having a subscriber model on `ConstraintType` and fields here `operator`? But yeah have to think about that longer.

Reply:
basically have your constraint type implement a method `evaluate_value` and/or `is_valid`.
From there LinearConstraint would just be
`is_valid(lc::LinearConstraint, xvalues) = a^T xvalues <= b` or something

If you need completely black-box functions, you can have
```julia
struct BBConstraint{F, T}
    func::F
    rhs::T
end
is_valid(f::BBFunction, xvalues) = f.func(xvalues) <= rhs
```

The only thing I see is that this is basically re-implementing lots of things from MOI, so it could be relevant to start from what they have

Reply:
The idea was that some constraints might share the same kind of constraint but are slightly different i.e have `==` vs. `<=` for the `LinearConstraint`. 
I prefer to have different functions for these cases instead of having `is_valid(lc, x)` and then check what the operator is in that function. If we have one `ConstraintType` for each constraint your idea would work but as some constraints share the same structure I found that wasteful.

Reply:
you can then replace what I put above with a method `evaluate_at_value(func, xvalues)` or something similar. Every function should be able to evaluate itself at given values. Then what happens with the value (either <=, >= or something else), is up to the constraint type

Reply:
Can you clarify what you mean by "evaluate"? 
Maybe the simple way of having a constraint.jl for each constraint works:
`LinearConstraint.jl`:
```
function prune!(com::CoM, lc::LinearConstraint)
      if lc.operator == :(==)
               return prune_eq_sum!(com, lc)
      end
      if lc.operator == :(<=)
               return prune_leq_sum!(com, lc)
     end
    ...
end
```
works and for the `BasicConstraint` we have an operator like `:alldifferent` or `:circuit` or something. If a new constraint gets added the developer needs to implement the normal prune functionality and then adds the `operator` to the constraint.jl
Then normal dispatch works + some `if` calls which should be okay performance wise, right?

Reply:
I was thinking of the case when you have a function and a right-hand-side.`f(x) operator rhs`. You can make your function a type you dispatch on